client/dashboard/sites/domains/primary-domain-selector.tsx

```diff
@@ -0,0 +1,195 @@
+	const primaryWithPlanOnly = userHasFlag( user, 'calypso_allow_nonprimary_domains_without_plan' );
+	const isOnFreePlan = site?.plan?.is_free ?? false;
+	const canUserSetPrimaryDomainOnThisSite =
+		! ( primaryWithPlanOnly && isOnFreePlan ) &&
+		( site?.plan?.features?.active.includes( 'set-primary-custom-domain' ) ?? false );
+	const setPrimaryDomainMutation = useMutation( siteSetPrimaryDomainMutation() );
+	const currentPrimaryDomain = domains.find( ( domain ) => domain.primary_domain )?.domain;
+	const domainsList = useMemo( () => {
+		if ( ! domains || ! site ) {
+			return [];
+		}
+		const hasWpcomStagingDomain = domains.find( ( domain ) => domain.is_wpcom_staging_domain );
+		return domains.filter( ( domain ) => {
+			// Basic eligibility criteria
+			const isEligible =
+				( domain.subtype.id === DomainSubtype.DOMAIN_REGISTRATION ||
+					domain.subtype.id === DomainSubtype.DOMAIN_CONNECTION ) &&
+				domain.can_set_as_primary &&
+				! domain.primary_domain &&
+				! domain.aftermarket_auction;
+
+			if ( ! isEligible ) {
+				return false;
+			}
+
+			// Additional filtering for wpcom staging domains
+			if ( hasWpcomStagingDomain && domain.wpcom_domain ) {
+				return domain.is_wpcom_staging_domain;
+			}
+
+			return true;
+		} );
+	}, [ domains, site ] );
+
+	if ( ! domains || ! site ) {
+		return null;
+	}
+
+	const fields: Field< { primaryDomain: string } >[] = [
+		{
+			id: 'primaryDomain',
+			label: __( 'Primary domain' ),
+			elements: [
+				{
+					label: __( 'Select a domain name' ),
+					value: '',
+				},
+				...domainsList.map( ( domain ) => {
+					return {
+						label: domain.domain,
+						value: domain.domain,
+					};
+				} ),
+			],
+		},
+	];
+
+	const form = {
+		layout: { type: 'regular' as const },
+		fields: [ 'primaryDomain' ],
+	};
+
+	const renderMessage = () => {
+		if ( ! canUserSetPrimaryDomainOnThisSite ) {
+			return createInterpolateElement(
+				'Your site plan doesn’t allow you to set a custom domain as a primary site address.<br/><upgradeLink>Upgrade to an annual paid plan</upgradeLink> and get a free one-year domain name registration or transfer. <learnMoreLink>Learn more</learnMoreLink>.',
+				{
+					upgradeLink: <a href={ `/plans/${ site.slug }` } />,
+					br: <br />,
+					learnMoreLink: <InlineSupportLink supportContext="primary-site-address" />,
+				}
+			);
+		}
+		if ( domainsList.length === 0 ) {
+			return createInterpolateElement(
+				'Before changing your primary site address you must register or connect a new custom domain. <learnMoreLink>Learn more</learnMoreLink>.',
+				{
+					learnMoreLink: <InlineSupportLink supportContext="primary-site-address" />,
+				}
+			);
+		}
+
+		return createInterpolateElement(
+			'The current primary site address for this site is: <currentPrimaryDomain />. This is the site address your visitors will see while browsing your site. <learnMoreLink>Learn more</learnMoreLink>.',
```

Comment:
Igor, sorry for not catching this earlier, but what do you think of updating the text to:

```suggestion
			'The current primary site address is: <currentPrimaryDomain />. This is the address your visitors will see while browsing your site. <learnMoreLink>Learn more</learnMoreLink>.',
```